pkg/bucket/bandwidth/measurements/bucket_measurement.go

```diff
@@ -0,0 +1,108 @@
+/*
+ * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package measurements
+
+import (
+	"math"
+	"sync"
+	"time"
+)
+
+const (
+	weight = 0.5
+)
+
+// BucketMeasurement captures the bandwidth details for one bucket
+type BucketMeasurement struct {
+	lock                 sync.Mutex
```

Comment:
> @kerneltime
> 
> I need to clarify some design stuff first. First, I need to understand the goals of this:
> 
> Our goal is to monitor bandwidth usage of replication on a per bucket basis and globally.
The Sum of all buckets should result in the global answer and the mc admin API will allow reading the values for the entire cluster.
> 
> Secondary goals:
> 
>     * The measurement should only have minimal impact on the system.
Yes
> 
>     * The measurements should be available to Prometheus?
Any one can read it. The initial use will be via Prometheus and mc admin 
> 
>     * Per object stats? To what degree?
Only the objects in flight. I have added the code but have not plumbed the admin API yet. This is to inspect a larger object replication. 
> 
>     * What is the granularity needed?
Per bucket across cluster and per server.
In flight objects across cluster and per server.
> 
>     * What is the acceptable lag? Meaning latency between the current time and the latest observed time.
Something that works for a human, a outage WAN outage or transient failure should be reflected in the measurements.
> 
> 
> In particular I am wondering if we need the detailed object tracking since it is quite heavy.
It is only for inflight objects, once done objects get cleared out unlike buckets. I initially had implemented an LRU for last n objects and measured bandwidth but I need to understand the mc admin experience before introducing it back.
The other part of the story is the bandwidth and concurrency throttling. So in flight replication objects should not be more then a few hundred thousand at peak per server.

Reply:
> It should be possible to use atomics since updates are additions and switching to next window can be done with a swap operation.
> 
> So if there is only one 'owner' of a bucket that updates the window this should be doable without a mutex.

Yup. Will update it